### DISCUSSION TITLE
View Only User with Edit Permissions

### DISCUSSION THREAD

**Discussion by hm-chris** [area17/twill#867] [Q&A]
I'm currently working on a feature request from within our company. It entails granting a View Only user permission to edit a single page. I've gotten a trait set up to handle saving and retrieving the users browser, which is working just fine. A problem I ran into is module edit links.

So, using pages as the module, when navigating to the pages list, the edit links are just text for View Only users. I found somewhat of a workaround by adding a check to the `getIndexOption` function in the `ModuleController`. It's a check to see if the option passed in is `'edit'`. If it is, just set `$authorized` to true. This seems awfully jank, and I don't know what repercussions this could have in other areas. 

Is there a more solid way to go about this? I noticed that edit links are available to View Only users in the recent activity section on the dashboard overview page, but I'm not quite able to figure out how to replicate that. 

Just to quickly clarify, a View Only user should still get a 403 when they click on an edit link for a page they weren't given edit permissions. I'm just trying to get the edit links to actually populate rather than simply be text.

**Comment by kirkbushell** [comment]
The permissions system inside Twill is really quite limited. I've been fighting it, trying to figure out how I can best support content contributors - users who can create and edit their own content but not see anything else. Currently, it seems impossible, and this is due to how how they've built the permissions checks, by not passing in the ids of the resources being modified.etc.

Would love to hear how you work this one out, if you do.

**Comment by kirkbushell** [comment]
So I think the way around this is to provide a PR that better supports permissions requirements, with Gates registered that support individual resources as well as the user. See: https://laravel.com/docs/8.x/authorization#writing-gates

Be keen to get some input from the mods about this, as I'd be happy to contribute this.

**Comment by antonioribeiro** [comment]
I agree a more granular permission system is something we need on Twill, but I believe we have something being cooked at AREA 17 right now about this, not sure how advanced this is because it's not on one of my projects, but maybe @ifox can? 

Sometimes we do client work but it's not quite ready for landing on Twill codebase, so if the work is partially done, maybe @kirkbushell can polish it for prime time? 😅

In the past I used [Bouncer](https://github.com/JosephSilber/bouncer) and found it simple and very powerful to use, as it integrates with Laravel Gates allowing us to check for permissions at Blade or Application levels, but if Vanilla Laravel Gates is better or would introduce less noise, I'm all for it too.

Cheers!

**Reply by kirkbushell** [reply]
I'd be happy to do this :)

**Reply by ifox** [reply]
We do have a very robust permissions system in the works. It allows groups, roles, and granular content/user permissions management, all in the CMS UI. It still needs some work to be released, but it is functional, we use it in production on our internal Twill instance that manages our website and internal tools. I think with enough help from the community we can get to a release pretty soon. I will regroup internally and open a PR asap.

**Reply by harveydobson** [reply]
@ifox  Do you guys have update on when you think the new permissions system might be available?

**Reply by ifox** [reply]
It will be available in 3.0 which will be released following the Laravel 9 release. See https://github.com/area17/twill/pull/1138

**Reply by harveydobson** [reply]
Very exciting!! Thanks very much @ifox

**Comment by ifox** [comment]
Hi there,

The purpose of view only users is to allow frontend previews of draft records through the CMS login. That's why links are disabled for them in the CMS, they are not really supposed to do anything in there.

What you are looking for would be best served by creating a new role with edit permissions but that you dynamically check like you seem to be already doing.

See https://twill.io/docs/#extending-user-roles-and-permissions.

**Comment by ifox** [comment]
Also, see https://github.com/area17/twill/discussions/912#discussioncomment-760860. I will share a repository with a complete setup later this week. There's a bit of cleaner way to do it than what you shared.
